### PR TITLE
Remove BinaryErrorHandler, rename SourceErrorHandler

### PR DESCRIPTION
**Pull request by binji**
Since SourceErrorHandler can support binary locations, it doesn't make
much sense to have two different error handlers. There is now only an
ErrorHandler base class, and the first argument specifies whether the
error handler is expecting text or binary locations.

### CODE REVIEW

File: src/error-handler.h [hunk 1]
```diff
@@ -24,14 +24,14 @@
...
 
-  virtual ~SourceErrorHandler() {}
+  virtual ~ErrorHandler() {}
 
   // Returns true if the error was handled.
-  virtual bool OnError(const Location*,
+  virtual bool OnError(const Location&,
                        const std::string& error,
                        const std::string& source_line,
                        size_t source_line_column_offset) = 0;
```
Comment (KarlSchimpf) [inline src/error-handler.h]: Why not introduce an inline version without source line and column, for use with the binary case. I.e. something like:

bool OnError(Offset offset, const std::string& error) {
  return OnError(Location(offset), error, std::string(), 0);
}

That way, the binary use doesn't need to explicitly convert/fill in arguments.

File: src/binary-reader-interpreter.cc [hunk 2]
```diff
@@ -322,7 +322,7 @@ Label* BinaryReaderInterpreter::TopLabel() {
 }
 
 bool BinaryReaderInterpreter::HandleError(Offset offset, const char* message) {
-  return error_handler->OnError(offset, message);
+  return error_handler->OnError(Location(offset), message, std::string(), 0);
```
Comment (KarlSchimpf) [inline src/binary-reader-interpreter.cc]: Use binary version of OnError (see error-handler.h comments).

File: src/binary-reader-ir.cc [hunk 3]
```diff
@@ -289,7 +289,7 @@ Result BinaryReaderIR::AppendExpr(Expr* expr) {
 }
 
 bool BinaryReaderIR::HandleError(Offset offset, const char* message) {
-  return error_handler->OnError(offset, message);
+  return error_handler->OnError(Location(offset), message, std::string(), 0);
```
Comment (KarlSchimpf) [inline src/binary-reader-ir.cc]: Again, use binary version of OnError.

File: src/error-handler.h (hunk 1 above)
Comment (binji) [inline src/error-handler.h]: Good idea.

File: src/common.h [hunk 4]
```diff
@@ -147,7 +147,8 @@ struct Location {
     Binary,
   };
 
-  Location() : filename(NULL), line(0), first_column(0), last_column(0) {}
+  Location() : filename(nullptr), line(0), first_column(0), last_column(0) {}
+  explicit Location(size_t offset) : filename(nullptr), offset(offset) {}
 
   const char* filename;
```
Comment (sbc100) [inline src/common.h]: Add "= nullptr" here to avoid setting in both contructors above?

File: test/dump/try-details.txt [hunk 5]
```diff
@@ -14,16 +14,21 @@
...
 Type:
  - type[0] () -> i32
 Function:
  - func[0] sig=0
 Custom:
  - name: "exception"
  - except[0] (i32)
+
 Code Disassembly:
+
```
Comment (sbc100) [inline test/dump/try-details.txt]: Strange.  I wonder how these newlines were originally missing?   Separate PR maybe?

Comment (binji) [inline test/dump/try-details.txt]: Yeah, I think they accidentally got included when I rebuilt all the tests. It is that same issue you found before, where empty lines are not caught by the diff.

File: src/common.h (hunk 4 above)
Comment (binji) [inline src/common.h]: Done.

**Pull request merged**